doc/tapscript_opcodes.md

```diff
@@ -0,0 +1,134 @@
+      13. Otherwise, fail.
+
+
+2. **Transaction Introspection codes**: Transaction introspection is already possible in elements script by use of `OP_CHECKSIGFROMSTACKVERIFY`, however the current solutions are really expensive in applications like [covenants](https://github.com/sanket1729/covenants-demo). Therefore, we are not adding any new functionality by supporting introspection, only making it easier to use. The warning still remains the same as with covenants, if the user is inspecting data from parts of the transaction that are not signed, the script can cause unexpected behavior. 
+   - Define `OP_SUCCESS201` as a multi-byte opcode `OP_INSPECTINPUT` with the following semantics. Execute the push opcode following `OP_INSPECTINPUT`, if the opcode is not a push opcode, fail. 
+      - Pop the stack pop as minimal `CScriptNum` as `n`. Next, pop another element as minimal `CScriptNum` input index `idx`.
+      1. If `n=0`, `OP_INSPECTINPUTSPENDTYPE` Push 1 byte "spend_type" onto the stack. spend_type (1) is equal to `(ext_flag * 2) + annex_present` as defined in [Modified BIP-341 SigMsg for Elements](https://gist.github.com/roconnor-blockstream/9f0753711153de2e254f7e54314f7169)
+      2. If `n=1`, `OP_INSPECTINPUTOUTPOINTFLAG` Push the outpoint_flag(1) as defined in [Modified BIP-341 SigMsg for Elements](https://gist.github.com/roconnor-blockstream/9f0753711153de2e254f7e54314f7169)
+      3. if `n=2`, `OP_INSPECTINPUTOUTPOINT` Push the outpoint as a tuple. First push the `txid`(32) of the `prev_out`, followed by a 4 byte push of `vout`
+      4. If `n=3`, `OP_INSPECTINPUTASSET` Push the `nAsset` as a tuple, first push the assetID(32), followed by the prefix(1)
+      5. If `n=4`, `OP_INSPECTINPUTVALUE` Push the `nValue` as a tuple, value(8, 32) followed by prefix,
+      6. If `n=5`, `OP_INSPECTINPUTSSCRIPTPUBKEY` Push the scriptPubkey(35) onto the stack.
+      7. If `n=6`, `OP_INSPECTINPUTSEQUENCE` Push the `nSequence`(4) as little-endian number. 
+      8. If `n=7`, `OP_INSPECTINPUTISSUANCE` Push the assetIssuance information(74-130) if the asset has issuance, otherwise push an empty vector
+      9. If `n=8`, `OP_INSPECTINPUTINDEX` Push the current input index(4) as little-endian onto the stack
+      10. If `n=9`, `OP_INSPECTINPUTANNEX` Push the annex onto the stack where the annex includes the prefix(0x50). If the annex does not exist, push an empty vector
+      11. Otherwise fail.
+   - Define `OP_SUCCESS202` as a multi-byte opcode `OP_INSPECTCURRENTINPUT` with the following semantics. Execute the push opcode following `OP_INSPECTCURRENTINPUT`, if the opcode is not a push opcode, fail. 
+      - Pop the stack pop as minimal `CScriptNum` as `n`. All the implementation is exactly the same as `OP_INSPECTINPUT`, but uses the current input instead of reading the input from stack. The names of the `OP_INSPECTINPUTx` would be replaced by `OP_INSPECTCURRENTINPUTx`
```

Comment:
```suggestion
      - Pop the stack as minimal `CScriptNum` as `n`. The implementation is exactly the same as `OP_INSPECTINPUT`, but uses the current input instead of reading the input from stack. The names of the `OP_INSPECTINPUTx` would be replaced by `OP_INSPECTCURRENTINPUTx`
```